pkg/snapshot/restorer/restorer.go

```diff
@@ -317,25 +326,192 @@ func startEmbeddedEtcd(ro RestoreOptions) (*embed.Etcd, error) {
+		fetcherIndex++
+		if fetcherIndex == numFetchers {
+			fetcherIndex = 0
+		}
+	}
+	return snapsPerFetcher, snapIndicesPerFetcher
+}
+
+// fetcher fetches the delta snapshots assigned to it by startFetchers, and runs as a goroutine
+func (r *Restorer) fetcher(snaps []snapstore.Snapshot, snapIndices []int, fetchedEventsInfo chan eventsInfo, errCh chan error) {
+	for i, snap := range snaps {
+		r.logger.Infof("Fetching delta snapshot %s", path.Join(snap.SnapDir, snap.SnapName))
+		eventsData, err := getEventsDataFromDeltaSnapshot(r.store, snap)
+		if err != nil {
+			errCh <- fmt.Errorf("failed to read events data from delta snapshot %s : %v", snap.SnapName, err)
+			break
+		}
+		r.logger.Infof("--- FETCHER - %s - fetched events data", snap.SnapName)
+		eventsInfo := eventsInfo{
+			Data:      eventsData,
+			Snap:      snap,
+			SnapIndex: snapIndices[i],
+		}
+		r.logger.Infof("--- FETCHER - %s - eventsInfo object created", snap.SnapName)
+		fetchedEventsInfo <- eventsInfo
+		r.logger.Infof("--- FETCHER - %s - fetchedEventsInfo data sent on channel", snap.SnapName)
+	}
+}
+
+func (r *Restorer) applier(client *clientv3.Client, remainingSnaps snapstore.SnapList, snapLocations []string, restoreFinished chan bool, errCh chan error) {
+	r.logger.Infof("--- APPLIER")
+
+	currSnapIndex := 0
+	for {
+		if snapLocations[currSnapIndex] != "" {
+			r.logger.Infof("--- APPLIER - snapshot %d found", currSnapIndex)
+			filePath := snapLocations[currSnapIndex]
+			currSnap := remainingSnaps[currSnapIndex]
+			eventsData, err := ioutil.ReadFile(filePath)
+			if err != nil {
+				errCh <- fmt.Errorf("failed to read events data from file for delta snapshot %s : %v", currSnap.SnapName, err)
+				break
+			}
+			defer os.Remove(filePath)
+
+			events := []event{}
+			err = json.Unmarshal(eventsData, &events)
+			if err != nil {
+				errCh <- fmt.Errorf("failed to read events from events data for delta snapshot %s : %v", currSnap.SnapName, err)
+				break
+			}
+
+			if err := applyEventsToEtcd(client, events); err != nil {
+				errCh <- fmt.Errorf("failed to apply events to etcd for delta snapshot %s : %v", currSnap.SnapName, err)
+				break
+			}
+			r.logger.Infof("--- APPLIER - applied events to etcd")
+
+			if err := verifySnapshotRevision(client, currSnap); err != nil {
+				errCh <- fmt.Errorf("snapshot revision verification failed for delta snapshot %s : %v", currSnap.SnapName, err)
+				break
+			}
+			r.logger.Infof("--- APPLIER - verified snapshot revision")
+
+			currSnapIndex++
+			r.logger.Infof("--- APPLIER - currSnapIndex incremented to %d", currSnapIndex)
+
+			if currSnapIndex == len(remainingSnaps) {
+				r.logger.Infof("--- APPLIER - currSnapIndex = len(remainingSnaps)")
+				restoreFinished <- true
+				r.logger.Infof("--- APPLIER - true sent on restoreFinished")
+				break
+			}
+		} else {
+			r.logger.Infof("--- APPLIER - waiting for snapIndex %d", currSnapIndex)
+			time.Sleep(waitDuration)
```

Comment:
In my original implementation, the goroutine `fetchListeners` updated the `snapLocations` array, and `applier` thread would poll this array to check for array updations, with a `waitDuration` to avoid hogging up CPU. But with the new worker pool implementation, a channel would be handling the synchronization.